### PR TITLE
 Added kernel_builder functions for the hardware efficient ansatz #361

### PR DESCRIPTION
**Pull request by johanneskuhlmann**
### Description

The hardware efficient ansatz was extended to allow kernel builder calling.
Additionally the hwe.h was refactored a bit to reduce code duplication via free function for defaults and forwarding.
Tests in python were added.

### CODE REVIEW

**Comment by johanneskuhlmann** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

**Comment by johanneskuhlmann** [comment]
recheck

File: python/runtime/cudaq/kernels/py_chemistry.cpp [hunk 1]
```diff
@@ -21,6 +22,36 @@ void bindChemistry(py::module &mod) {
           "kernel. This function takes as input the existing `cudaq.Kernel` to "
           "append to, pre-allocated qubits, list of parameters, number of "
           "electrons, and number of qubits as input, in that order.");
+
+  mod.def("num_hwe_parameters", &cudaq::num_hwe_parameters,
+          "For the given number of qubits and layers, return the required "
+          "number of hwe parameters.");
+  // use explicit overload as we have two overloaded signatures, i.e. with and
+  // without CNOT couplers
+  //   mod.def("hwe_cnot", &cudaq::hwe_cnot<kernel_builder<>>,
```
Comment (amccaskey) [inline python/runtime/cudaq/kernels/py_chemistry.cpp:31]: Would you mind removing this comment of old-code?

**Review by amccaskey — APPROVED** [review]
Overall, looks great. Thanks for the contribution!

File: python/runtime/cudaq/kernels/py_chemistry.cpp (hunk 1 above)
Comment (johanneskuhlmann) [reply python/runtime/cudaq/kernels/py_chemistry.cpp]: Removed commented out code line.

**Pull request merged**